skbio/tree/tests/test_tree.py

```diff
@@ -1303,6 +1303,51 @@ def test_shuffle_raises(self):
         with self.assertRaises(MissingNodeError):
             next(self.simple_t.shuffle(names=['x', 'y']))
 
+    def test_cache_leafcount(self):
+        t = TreeNode.read([r'((a,b)e,(c,d)f,g)r;'])
+
+        self.assertEquals(t.find('r').leafcount, 5)
+        self.assertEquals(t.find('e').leafcount, 2)
+        self.assertEquals(t.find('f').leafcount, 2)
+        self.assertEquals(t.find('a').leafcount, 1)
+        self.assertEquals(t.find('b').leafcount, 1)
+        self.assertEquals(t.find('c').leafcount, 1)
+        self.assertEquals(t.find('d').leafcount, 1)
+        self.assertEquals(t.find('g').leafcount, 1)
+        self.assertEquals(t.find('r').leafcount, 5)
+
+    def test_cache_height(self):
+        t = TreeNode.read(io.StringIO("((a:2,b:6)c:4,(d:5,e:1)f:7)r:9;"))
+
+        self.assertEquals(t.find('r').height, 21.0)
+        self.assertEquals(t.find('f').height, 12.0)
+        self.assertEquals(t.find('c').height, 10.0)
+        self.assertEquals(t.find('d').height, 5.0)
+        self.assertEquals(t.find('e').height, 1.0)
+        self.assertEquals(t.find('b').height, 6.0)
+        self.assertEquals(t.find('a').height, 2.0)
+
+    def test_cache_height_error(self):
+        t = TreeNode.read(io.StringIO("((a:2,b:6)c:4,(d:5,e)f:7)r:9;"))
+        with self.assertRaises(ValueError):
+            t.height
```

Comment:
This exception will be thrown if the root does not have a length as well given how the code is written

I can see why it would be nice to implicitly treat length as zero, or to except. I'd prefer to get a second opinion on that